src/HexagonOptimize.cpp

```diff
@@ -1804,6 +1804,88 @@ class VtmpyGenerator : public IRMutator {
         IRMutator::visit(op);
     }
 };
+
+// Convert some expressions to an equivalent form which could get better
+// optimized in later stages for hexagon
+class RearrangeExpressions : public IRMutator {
+private:
+    using IRMutator::visit;
+
+    void visit(const Mul *op) {
+        if (op->type.is_vector()) {
+            if (op->a.as<Broadcast>()) {
+                // Ensures broadcasts always occurs as op1 not op0
+                expr = mutate(op->b * op->a);
+                return;
+            } else if (op->b.as<Broadcast>()) {
+                // Distributing broadcasts helps creating more vmpa
+                // because of more adds of muls. Since muls are
+                // generally widening ops we need not check if op->a
+                // is a sum of widening casts.
+                vector<Expr> matches;
+                static const vector<Expr> add_patterns = {
+                    (wild_i16x + wild_i16x),
+                    (wild_i32x + wild_i32x),
+                };
+                for (const Expr &p : add_patterns) {
+                    if(expr_match(p, op->a, matches)) {
+                        // simplify() ensures that if matches[0] is also
+                        // a scaler multiplications, then we combine the two
+                        // broadcasts produced in matches[0] * op->b. For eg:
+                        // matches[0] = bc * wild_i16x, then simplify combines
+                        // bc * op->b into a single expression.
+                        expr = mutate(simplify(matches[0] * op->b) +
```

Comment:
I think it's dangerous to be calling simplify in this mutator, since these rewrites are possibly going to fight with the rewrites in the simplifier, leading to infinite recursion. If nothing else, it might undo what you are trying to do here.